orchestrator/careplancontributor/ehr/kafka_client.go

```diff
@@ -0,0 +1,169 @@
+// It holds the Kafka topic and producer used to submit messages.
+type KafkaClientImpl struct {
+	topic    string
+	producer *kafka.Producer
+}
+
+type NoopClientImpl struct {
+}
+
+// NewClient creates a new KafkaClient based on the provided KafkaConfig.
+// If Kafka is enabled in the configuration, it initializes a Kafka producer
+// and sets up a delivery report handler for produced messages. If Kafka is
+// not enabled, it returns a NoopClientImpl which writes messages to a file.
+//
+// Parameters:
+//   - config: KafkaConfig containing the configuration for the Kafka client.
+//
+// Returns:
+//   - KafkaClient: An implementation of the KafkaClient interface.
+//   - error: An error if the Kafka producer could not be created.
+func NewClient(config KafkaConfig) (KafkaClient, error) {
+
+	var kafkaClient KafkaClient
+	if config.Enabled {
+		endpoint := config.Endpoint
+		log.Infof("KafkaClientImpl, connecting to %s, Mechanism: %s, protocol: %s, username: %s, password: %s",
+			endpoint,
+			config.Sasl.Mechanism,
+			config.Security.Protocol,
+			config.Sasl.Username,
+			config.Sasl.Password)
+		producer, err := kafka.NewProducer(&kafka.ConfigMap{
+			"bootstrap.servers": endpoint,
+			"sasl.mechanisms":   config.Sasl.Mechanism,
+			"sasl.username":     config.Sasl.Username,
+			"sasl.password":     config.Sasl.Password,
+			"security.protocol": config.Security.Protocol,
+		})
+		if err != nil {
+			return nil, err
+		}
+
+		// Delivery report handler for produced messages
+		go func() {
+			log.Infof("Kafka func started")
+			for e := range producer.Events() {
+				log.Infof("Kafka event received: %v", e.String())
+				switch ev := e.(type) {
+				case *kafka.Message:
+					if ev.TopicPartition.Error != nil {
+						log.Infof("Kafka delivery failed: %v\n", ev.TopicPartition)
+					} else {
+						log.Errorf("Kafka delivered message to %v\n", ev.TopicPartition)
+					}
+				}
+			}
+			log.Infof("Kafka func ended")
+		}()
+
+		kafkaClient = &KafkaClientImpl{
+			topic:    config.Topic,
+			producer: producer,
+		}
+	} else {
+		kafkaClient = &NoopClientImpl{}
+	}
+	return kafkaClient, nil
+}
+
+// SubmitMessage submits a message to the Kafka topic associated with the KafkaClientImpl.
+// It produces a message with the given key and value to the Kafka producer.
+//
+// Parameters:
+//   - key: The key of the message to be submitted.
+//   - value: The value of the message to be submitted.
+//
+// Returns:
+//   - error: An error if the message could not be produced.
+func (k *KafkaClientImpl) SubmitMessage(key string, value string) error {
+	log.Infof("SubmitMessage, submitting key %s", key)
+	err := k.producer.Produce(&kafka.Message{
+		TopicPartition: kafka.TopicPartition{
+			Topic:     &k.topic,
+			Partition: kafka.PartitionAny,
+		},
+		Key:   []byte(key),
+		Value: []byte(value),
+	}, nil)
+	if err != nil {
+		return fmt.Errorf("failed to write message: %w", err)
+	}
+	k.producer.Flush(10 * 1000)
```

Comment:
why is this and what is that value?

Reply:
I think it is good practice to flush the connection before continuing.